### ISSUE TITLE
jasmine-client-integration tests only run when app physically launched in browser

### ISSUE THREAD

**Issue by philipobrien** [Sanjo/meteor-jasmine#266]
If I have no browser window open and run `meteor --test` my `jasmine-server-unit` tests run but then terminal output stalls when I expect the `jasmine-client-integration` tests to run. The logs show nothing, the most recent log being `[velocity] Established connection with Velocity.`

If I then open the running app in my browser the client-integration tests complete successfully. How do I get around this? 

I'm using Meteor 1.1.0.3 and the package versions I am using are

```
accounts-base@1.2.0
accounts-password@1.1.1
alanning:roles@1.2.13
aldeed:template-extension@3.4.3
amplify@1.0.0
autoupdate@1.2.1
babrahams:editable-json@0.4.3
base64@1.0.3
binary-heap@1.0.3
blaze@2.1.2
blaze-tools@1.0.3
boilerplate-generator@1.0.3
callback-hook@1.0.3
check@1.0.5
coffeescript@1.0.6
dburles:google-maps@1.1.3
dburles:mongo-collection-instances@0.3.4
ddp@1.1.0
deps@1.0.7
ejson@1.0.6
email@1.0.6
fastclick@1.0.3
fortawesome:fontawesome@4.4.0
geojson-utils@1.0.3
grigio:babel@0.1.6
gwendall:session-json@0.1.7
html-tools@1.0.4
htmljs@1.0.4
http@1.1.0
id-map@1.0.3
imsky:holder@2.8.1
insecure@1.0.3
iron:controller@1.0.8
iron:core@1.0.8
iron:dynamic-template@1.0.8
iron:layout@1.0.8
iron:location@1.0.9
iron:middleware-stack@1.0.9
iron:router@1.0.9
iron:url@1.0.9
jquery@1.11.3_2
json@1.0.3
lai:collection-extensions@0.1.4
launch-screen@1.0.2
less@1.0.14
livedata@1.0.13
localstorage@1.0.3
logging@1.0.7
manuelschoebel:wait-on-lib@0.3.0
materialize:materialize@0.97.0
matteodem:easy-search@1.6.2
meteor@1.1.6
meteor-platform@1.2.2
meteorhacks:aggregate@1.2.1
meteorhacks:collection-utils@1.2.0
meteorhacks:kadira@2.23.0
meteorhacks:meteorx@1.3.1
meteortoys:toykit@0.8.7
minifiers@1.1.5
minimongo@1.0.8
mobile-status-bar@1.0.3
mongo@1.1.0
mongo-livedata@1.0.8
msavin:mongol@1.1.5
nicolaslopezj:tabular-materialize@1.2.1
npm-bcrypt@0.7.8_2
observe-sequence@1.0.6
ordered-dict@1.0.3
package-version-parser@3.0.3
peernohell:c3@1.1.3
practicalmeteor:chai@2.1.0_1
practicalmeteor:loglevel@1.2.0_2
random@1.0.3
reactive-dict@1.1.0
reactive-var@1.0.5
reload@1.1.3
retry@1.0.3
routepolicy@1.0.5
sacha:spin@2.3.1
sanjo:jasmine@0.17.0
sanjo:karma@1.6.1
sanjo:long-running-child-process@1.1.3
sanjo:meteor-files-helpers@1.1.0_7
sanjo:meteor-version@1.0.0
sergeyt:d3@3.4.1
service-configuration@1.0.4
session@1.1.0
sha@1.0.3
softwarerero:accounts-t9n@1.0.9
spacebars@1.0.6
spacebars-compiler@1.0.6
srp@1.0.3
stevezhu:lodash@3.10.1
templating@1.1.1
tracker@1.0.7
ui@1.0.6
underscore@1.0.3
url@1.0.4
useraccounts:core@1.12.1
useraccounts:iron-routing@1.12.1
useraccounts:materialize@1.12.1
velocity:chokidar@1.0.3_1
velocity:core@0.9.3
velocity:html-reporter@0.8.2
velocity:meteor-internals@1.1.0_7
velocity:meteor-stubs@1.1.0
velocity:shim@0.1.0
velocity:source-map-support@0.3.2_1
webapp@1.2.0
webapp-hashing@1.0.3
```

**Comment by philipobrien** [comment]
Another update, I am not getting some output from PhantomJS but only

```
[PhantomJS] TypeError: instanceof called on an object with an invalid prototype property.
```

at which point nothing happens unless I open the app in my browser and then the tests complete

**Comment by Sanjo** [comment]
I don't have a clear solution but two things that may help:

* Try to use [PhantomJS 2](http://phantomjs.org/download.html)
* Try to add es5-shim to your app (not sure if that works for Meteor 1.1.0.2 apps). https://atmospherejs.com/meteor/es5-shim

**Comment by philipobrien** [comment]
Thanks, I just tried both of those but still no luck I'm afraid

**Comment by ubald** [comment]
Same thing here, I run:
```
JASMINE_BROWSER=PhantomJS velocity test-package --driver-package velocity:console-reporter my:package
```
and it doesn't run in PhantomJS, instead it says:
```
=> App running at: http://localhost:3000/
```
and I have to go in the browser to run the tests.

**Comment by philipobrien** [comment]
If I move the `client-integration` tests to the `client-unit` directory and remove the `client-integration` directory altogether, then the tests run with the command

```
PHANTOMJS_BIN=/usr/local/share/phantomjs/phantomjs JASMINE_BROWSER=PhantomJS meteor --test
```

Are there are any problems going this route?

**Comment by Sanjo** [comment]
@philipobrien No, no problems. If that works for you, great!

I just realized that you cannot use PhantomJS 2 for `velocity ... --ci` right now. I've created an issue here: https://github.com/meteor-velocity/meteor/issues/4.

@ubald The JASMINE_BROWSER setting does only affect the client unit mode, currently. The client integration tests always run in PhantomJs right now on CI.

**Comment by philipobrien** [comment]
@Sanjo I should add that when  I run 

```
PHANTOMJS_BIN=/usr/local/share/phantomjs/phantomjs JASMINE_BROWSER=PhantomJS meteor --test
```

it is the PhantomJS 2 binary I am pointing to, and it works as long as all the client tests reside under the `client-unit` directory

**Comment by Sanjo** [comment]
@philipobrien Yeah, got it. In the client unit mode you can also run your tests in Chrome or Firefox in CI. You just need to change the JASMINE_BROWSER setting.

**Comment by ubald** [comment]
@Sanjo What if I want to run tests in PhantomJS not in CI, so that I can only have a terminal opened showing me the results?

**Comment by Sanjo** [comment]
@ubald You can do that with the client unit mode (JASMINE_BROWSER=PhantomJS). But not with the client integration mode currently.

**Comment by philipobrien** [comment]
This seemed to be working for me but it has started throwing errors again

```console
[PhantomJS] TypeError: instanceof called on an object with an invalid prototype property.
...
...
[PhantomJS] ReferenceError: Can't find variable: Template
```
This is when being run by Jenkins with the following command

```console
VELOCITY_MONGO_URL=xxx PHANTOMJS_BIN=/usr/local/bin/phantomjs JASMINE_BROWSER=PhantomJS \ 
/usr/local/bin/meteor --test --release velocity:METEOR@1.1.0.3_1;
```

so is it not possible to do this with CI?

**Comment by abernix** [comment]
@philipobrien, I have CI working on a rather decent size project, using PhantomJS, on both Semaphore CI and CodeShip.  I was also having problems with *only* `jasmine-client-integration` tests not running (the mirror would appear to get initialized by Velocity, but then the tests would never run and no error was outputted in the logs).

I ended up rearranging my `.meteor/packages` and it now works for me very reliably.  I never fully got to the bottom of what was happening but I think it had something to do with modules that sub-class or change `Meteor.Collection` or `Mongo.Collection`.

Can you reveal the order that your packages are listed in your `.meteor/packages` file within your project root?

Also, not that it will fix the problem for you, but `velocity:METEOR@1.1.0.3_2` (note `_2`) is out now, which might help additionally - I know I definitely had problems using the official meteor release.

**Comment by philipobrien** [comment]
@abernix here is the order

```
meteor-platform
jquery
iron:router
peernohell:c3
sanjo:jasmine
sacha:spin
accounts-password
dburles:google-maps
imsky:holder
alanning:roles
msavin:mongol
fortawesome:fontawesome
manuelschoebel:wait-on-lib
materialize:materialize
nicolaslopezj:tabular-materialize
useraccounts:materialize
meteorhacks:kadira
matteodem:easy-search
reactive-var
velocity:html-reporter
stevezhu:lodash
useraccounts:iron-routing
es5-shim
xolvio:cucumber
```

**Comment by abernix** [comment]
Cool. I might recommend a couple changes... Having had *multiple* issues with package order screwing up my builds, I'm a bit picky about the order — your mileage may vary, but might be worth a shot. 

* generally try to keep "core" packages (ones that do not include a colon `:` in them), together at the top. 
* move all test framework packages directly after the "core" packages
* putting `mongol` as far down as possible.  The bottom might be best but definitely after the test frameworks.
* putting `velocity:*` stuff before other framework stuff. 
* putting your `iron-router` router after velocity.

Basically, this helps if any packages screw up prototypes, it will happen later in the process with less complex packages which are less likely to be affected.  It happens, trust me.

**Comment by philipobrien** [comment]
OK, I changed it to this

```
meteor-platform
jquery
accounts-password
iron:router
peernohell:c3
sanjo:jasmine
sacha:spin
reactive-var
es5-shim
xolvio:cucumber
velocity:html-reporter
useraccounts:iron-routing
dburles:google-maps
imsky:holder
alanning:roles
fortawesome:fontawesome
manuelschoebel:wait-on-lib
materialize:materialize
nicolaslopezj:tabular-materialize
useraccounts:materialize
meteorhacks:kadira
matteodem:easy-search
stevezhu:lodash
msavin:mongol
```

Unfortunately it hasn't fixed things. In fact, on my local machine I am now getting similar Velocity errors

```
[PhantomJS] TypeError: instanceof called on an object with an invalid prototype property.
...
[PhantomJS]   http://localhost:59364/packages/global-imports.js?463bf9088a082f70cace700ebe7ff66c60248da1:8
[PhantomJS] ReferenceError: Can't find variable: Template
...
...
```

yet despite these errors all Cucumber tests run and it ends with `TESTS RAN SUCCESSFULLY`

**Comment by philipobrien** [comment]
Hmmmm, my cucumber log is showing this for the first test

```
Scenario: An adviser can login with valid information   [2m[37m# features/adviser/adviser_login.feature:11[0m
    [31mGiven The adviser is logged out[0m                       [2m[37m# features/adviser/adviser_login.feature:8[0m
[33m      Error: POST /session//url[39m
[33m      Build info: version: '2.46.0', revision: '87c69e2', time: '2015-06-04 16:16:47'[39m
[33m      System info: host: 'smart-appi-jenkins', ip: '192.168.1.6', os.name: 'Linux', os.arch: 'amd64', os.version: '3.13.0-53-generic', java.version: '1.7.0_79'[39m
[33m      Driver info: driver.version: unknown[39m
[35m[39m
[35m[39m
[31mRuntimeError: POST /session//url
```

**Comment by abernix** [comment]
Well, I like the idea that things are at least being more similar!  Heh.

Maybe play with the order a bit more and see if you can work through some different results.  I played with mine for hours and eventually found that I absolutely had to have certain packages after others.  Some packages just break things for others - while I despise the way that these packages do what they do, some were critical to my workflow and I just had to keep them in.  Try adding/removing packages where you can and see about different results.

Also, if you're working on your local machine, maybe try to use Chrome instead of PhantomJS for a minute and debug in the browsers themselves.

**Comment by philipobrien** [comment]
There are certainly issues with Jasmine but in case it's more a fault of Velocity I have created an issue [here](https://github.com/meteor-velocity/velocity/issues/371). I have also started a repo [here](https://github.com/philipobrien/velocity-debug) for the sole purpose of working through these issues (and hopefully solving them!)

**Comment by abernix** [comment]
Thanks for the demo repo -- life-saver! :kissing_closed_eyes: 

I have posted my findings in the [other ticket](/meteor-velocity/velocity/issues/371).  If I pass CUCUMBER=0, the problem goes away.  Also, if we remove `xolvio:cucumber`, the problem goes away as well.

I can't confirm now, as I've got my project running successfully with both, but I did have a point and time when `jasmine-client-integration` tests weren't running.  I resolved it by re-ordering my `packages` load order, but I wonder now if disabling Cucumber would have fixed it.  I will check if I get a chance.
